# AstraZeneca Coding Assignment 

## Done by Ganesh Prasanth Arumuganainar

## 

Priting the date and time stamp just to keep track of the time spent on the solution

In [1]:
import datetime
now = datetime.datetime.now()
print ("Starting date and time : ")
print (now.strftime("%Y-%m-%d %H:%M:%S"))

Starting date and time : 
2021-04-16 12:11:32


## Part A - Question

Using the data from the OpenFDA API (documentation at https://open.fda.gov/apis/drug/label/) to determine the average number of ingredients (spl_product_data_elements) contained in AstraZeneca medicines per year. 

The output should look similar to:
<pre>
year   drug_names     avg_number_of_ingredients
2018   drugA,drugB    21
... 
<\pre>

## Part A - Solution

### Importing and exploring data

In [2]:
# Importing necessary packages

import pandas as pd
import json
import requests
from pandas.io.json import json_normalize
import matplotlib.pyplot as plt

The API call for AstraZeneca data is 'https://api.fda.gov/drug/label.json?search=openfda.manufacturer_name:"AstraZeneca"'

In [3]:
# Get the JSON response using API request (limit = 1)

response = requests.get('https://api.fda.gov/drug/label.json?search=openfda.manufacturer_name:"AstraZeneca"')
json_response = response.json()
json_response.keys()

dict_keys(['meta', 'results'])

In [4]:
json_response.get('meta')

{'disclaimer': 'Do not rely on openFDA to make decisions regarding medical care. While we make every effort to ensure that data is accurate, you should assume all results are unvalidated. We may limit or otherwise restrict your access to the API in line with our Terms of Service.',
 'terms': 'https://open.fda.gov/terms/',
 'license': 'https://open.fda.gov/license/',
 'last_updated': '2021-04-15',
 'results': {'skip': 0, 'limit': 1, 'total': 37}}

There are 37 results in total under the manufacturer name : AstraZeneca. 
So, the limit could be changed to 37 instead of default 1. 
The API call for all 37 AstraZeneca data is 'https://api.fda.gov/drug/label.json?search=openfda.manufacturer_name:"AstraZeneca"&limit=37'

In [5]:
# Get the JSON response using API request (limit = 37)

response = requests.get('https://api.fda.gov/drug/label.json?search=openfda.manufacturer_name:"AstraZeneca"&limit=37')
json_response = response.json()
json_response.keys()

dict_keys(['meta', 'results'])

In [6]:
json_response.get('meta')

{'disclaimer': 'Do not rely on openFDA to make decisions regarding medical care. While we make every effort to ensure that data is accurate, you should assume all results are unvalidated. We may limit or otherwise restrict your access to the API in line with our Terms of Service.',
 'terms': 'https://open.fda.gov/terms/',
 'license': 'https://open.fda.gov/license/',
 'last_updated': '2021-04-15',
 'results': {'skip': 0, 'limit': 37, 'total': 37}}

In [7]:
# Normalise the JSON data into Pandas dataframes

df = json_normalize(json_response.get('results'))
df.head()

/opt/anaconda3/envs/python-cvcourse/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


effective_time                               recent_major_changes  \
0       20200918        [Warnings and Precautions ( 5.20 ) 09/2020]   
1       20201005                                                NaN   
2       20190724                                                NaN   
3       20201202  [Contraindications ( 4 ) 2/2020 Warnings and P...   
4       20191011                                                NaN   

                                   drug_interactions  \
0  [7 DRUG INTERACTIONS • Concomitant use of stro...   
1  [7 DRUG INTERACTIONS There are no known drug-d...   
2  [7 DRUG INTERACTIONS In clinical studies, conc...   
3  [7 DRUG INTERACTIONS Table 3: Clinically Relev...   
4  [7 DRUG INTERACTIONS Strong Cytochrome P450 3A...   

                                       geriatric_use  \
0  [8.5 Geriatric Use Of the approximately 3700 p...   
1  [8.5 Geriatric Use For fulvestrant 250 mg, whe...   
2  [8.5 Geriatric Use Of the total number of asth...   
3  [8.5 Geriatric Use In two comparator-controlle...   
4  [8.5 Geriatric Use Of the 215 patients in 3 cl...   

                                               abuse  \
0  [9.2 Abuse SEROQUEL has not been systematicall...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                    pharmacodynamics  \
0  [12.2 Pharmacodynamics Quetiapine and its meta...   
1  [12.2 Pharmacodynamics In a clinical study in ...   
2  [12.2 Pharmacodynamics Asthma Cardiovascular e...   
3  [12.2 Pharmacodynamics Exenatide improves glyc...   
4  [12.2 Pharmacodynamics The therapeutic effects...   

                                         description  \
0  [11 DESCRIPTION SEROQUEL ® (quetiapine) is an ...   
1  [11 DESCRIPTION FULVESTRANT INJECTION for intr...   
2  [11 DESCRIPTION BUDESONIDE AND FORMOTEROL FUMA...   
3  [11 DESCRIPTION BYDUREON BCISE (exenatide exte...   
4  [11 DESCRIPTION Budesonide, the active compone...   

                              nonclinical_toxicology  \
0  [13 NONCLINICAL TOXICOLOGY 13.1 Carcinogenesis...   
1  [13 NONCLINICAL TOXICOLOGY 13.1 Carcinogenesis...   
2  [13 NONCLINICAL TOXICOLOGY 13.1 Carcinogenesis...   
3  [13 NONCLINICAL TOXICOLOGY 13.1 Carcinogenesis...   
4  [13 NONCLINICAL TOXICOLOGY 13.1 Carcinogenesis...   

                         warnings_and_cautions_table  \
0  [<table width="100%"><caption>Table 2: Drug-Pl...   
1                                                NaN   
2  [<table ID="_RefID0EKJAE" width="100%"> <capti...   
3                                                NaN   
4                                                NaN   

                          dosage_forms_and_strengths  ...  \
0  [3 DOSAGE FORMS AND STRENGTHS • 25 mg tablets ...  ...   
1  [3 DOSAGE FORMS AND STRENGTHS FULVESTRANT INJE...  ...   
2  [3 DOSAGE FORMS AND STRENGTHS BUDESONIDE AND F...  ...   
3  [3 DOSAGE FORMS AND STRENGTHS Extended-release...  ...   
4  [3 DOSAGE FORMS AND STRENGTHS PULMICORT RESPUL...  ...   

  storage_and_handling_table patient_medication_information  \
0                        NaN                            NaN   
1                        NaN                            NaN   
2                        NaN                            NaN   
3                        NaN                            NaN   
4                        NaN                            NaN   

  patient_medication_information_table risks spl_unclassified_section  \
0                                  NaN   NaN                      NaN   
1                                  NaN   NaN                      NaN   
2                                  NaN   NaN                      NaN   
3                                  NaN   NaN                      NaN   
4                                  NaN   NaN                      NaN   

  laboratory_tests package_labe

### Extracting year from 'effective_time' 

In [8]:
df['year'] = pd.DatetimeIndex(df['effective_time']).year

### Extracting only the columns of interest

In [12]:
df = df[['year','spl_product_data_elements', 'openfda.generic_name']]
df.head()

year                          spl_product_data_elements  \
0  2020  [SEROQUEL quetiapine QUETIAPINE FUMARATE QUETI...   
1  2020  [Fulvestrant Fulvestrant FULVESTRANT FULVESTRA...   
2  2019  [Budesonide and Formoterol Fumarate Dihydrate ...   
3  2020  [BYDUREON BCise exenatide EXENATIDE EXENATIDE ...   
4  2019  [PULMICORT RESPULES Budesonide BUDESONIDE BUDE...   

                             openfda.generic_name  
0                                    [QUETIAPINE]  
1                                   [FULVESTRANT]  
2  [BUDESONIDE AND FORMOTEROL FUMARATE DIHYDRATE]  
3                                     [EXENATIDE]  
4                                    [BUDESONIDE]

### Inspecting spl_product_data_elements

In [14]:
df['spl_product_data_elements'][0]

['SEROQUEL quetiapine QUETIAPINE FUMARATE QUETIAPINE POVIDONE, UNSPECIFIED DIBASIC CALCIUM PHOSPHATE DIHYDRATE MICROCRYSTALLINE CELLULOSE SODIUM STARCH GLYCOLATE TYPE A POTATO MAGNESIUM STEARATE HYPROMELLOSE, UNSPECIFIED POLYETHYLENE GLYCOL, UNSPECIFIED TITANIUM DIOXIDE FERRIC OXIDE RED FERRIC OXIDE YELLOW LACTOSE MONOHYDRATE WATER peach biconvex SEROQUEL;25 SEROQUEL quetiapine QUETIAPINE FUMARATE QUETIAPINE POVIDONE, UNSPECIFIED DIBASIC CALCIUM PHOSPHATE DIHYDRATE MICROCRYSTALLINE CELLULOSE SODIUM STARCH GLYCOLATE TYPE A POTATO MAGNESIUM STEARATE HYPROMELLOSE, UNSPECIFIED POLYETHYLENE GLYCOL, UNSPECIFIED TITANIUM DIOXIDE LACTOSE MONOHYDRATE WATER FERRIC OXIDE YELLOW biconvex SEROQUEL;50 SEROQUEL quetiapine QUETIAPINE FUMARATE QUETIAPINE POVIDONE, UNSPECIFIED DIBASIC CALCIUM PHOSPHATE DIHYDRATE MICROCRYSTALLINE CELLULOSE SODIUM STARCH GLYCOLATE TYPE A POTATO LACTOSE MONOHYDRATE MAGNESIUM STEARATE HYPROMELLOSE, UNSPECIFIED POLYETHYLENE GLYCOL, UNSPECIFIED TITANIUM DIOXIDE FERRIC OXIDE Y

### Get the number of ingredients

In [15]:
# Since each spl_product_data_element is separated from each other by commas, we split them and get the number of elements

df['number_of_ingredients'] = df['spl_product_data_elements'].apply(lambda x: len(x[0].split(',')))
df.head()

year                          spl_product_data_elements  \
0  2020  [SEROQUEL quetiapine QUETIAPINE FUMARATE QUETI...   
1  2020  [Fulvestrant Fulvestrant FULVESTRANT FULVESTRA...   
2  2019  [Budesonide and Formoterol Fumarate Dihydrate ...   
3  2020  [BYDUREON BCise exenatide EXENATIDE EXENATIDE ...   
4  2019  [PULMICORT RESPULES Budesonide BUDESONIDE BUDE...   

                             openfda.generic_name  number_of_ingredients  
0                                    [QUETIAPINE]                     19  
1                                   [FULVESTRANT]                      1  
2  [BUDESONIDE AND FORMOTEROL FUMARATE DIHYDRATE]                      1  
3                                     [EXENATIDE]                      7  
4                                    [BUDESONIDE]                      4

### Grouping the data by year, summing up the drug names, averaging the number of ingredients

In [18]:
result1 = df.groupby('year', as_index=False).agg({'openfda.generic_name':'sum','number_of_ingredients':'mean'})
result1.head()

year                               openfda.generic_name  \
0  2016                             [METOPROLOL SUCCINATE]   
1  2018                           [ESOMEPRAZOLE MAGNESIUM]   
2  2019  [BUDESONIDE AND FORMOTEROL FUMARATE DIHYDRATE,...   
3  2020  [QUETIAPINE, FULVESTRANT, EXENATIDE, DAPAGLIFL...   
4  2021  [OLAPARIB, DURVALUMAB, ACLIDINIUM BROMIDE, BEN...   

   number_of_ingredients  
0                    1.0  
1                    9.0  
2                    2.4  
3                    6.0  
4                    1.8

### Customising the column names as per requirements

In [19]:
result1.columns = ['year','drug_names','avg_number_of_ingredients']
result1

year                                         drug_names  \
0  2016                             [METOPROLOL SUCCINATE]   
1  2018                           [ESOMEPRAZOLE MAGNESIUM]   
2  2019  [BUDESONIDE AND FORMOTEROL FUMARATE DIHYDRATE,...   
3  2020  [QUETIAPINE, FULVESTRANT, EXENATIDE, DAPAGLIFL...   
4  2021  [OLAPARIB, DURVALUMAB, ACLIDINIUM BROMIDE, BEN...   

   avg_number_of_ingredients  
0                        1.0  
1                        9.0  
2                        2.4  
3                        6.0  
4                        1.8

### Insights

On average, AstraZeneca used the highest number of ingredients in 2018 and the lowest number of ingredients in 2016

## Part B - Question

Repeat the same analysis, calculate the average number of ingredients per year and per delivery route for all manufacturers.

The output should look similar to:

<pre>
year   route      avg_number_of_ingredients
2018   oral       123
2018   injection  213
2017   ...        ...
...
<\pre>

## Part B - Solution

### Importing and exploring data

The API call for the entire dataset is 'https://api.fda.gov/drug/label.json?'

In [32]:
# Get the JSON response using API request (limit = 1)

response = requests.get('https://api.fda.gov/drug/label.json?')
json_response = response.json()
json_response.keys()

dict_keys(['meta', 'results'])

In [33]:
json_response['meta']

{'disclaimer': 'Do not rely on openFDA to make decisions regarding medical care. While we make every effort to ensure that data is accurate, you should assume all results are unvalidated. We may limit or otherwise restrict your access to the API in line with our Terms of Service.',
 'terms': 'https://open.fda.gov/terms/',
 'license': 'https://open.fda.gov/license/',
 'last_updated': '2021-04-15',
 'results': {'skip': 0, 'limit': 1, 'total': 185324}}

### Assumption

There are 185324 results available but each API request shall be responded with a limit of 1000 data only. In case of importing the entire dataset, more API requests need to be fired and the process is time consuming. For convenience, we shall proceed with 1000. 

In [34]:
# Get the JSON response using API request (limit = 1000)

response = requests.get('https://api.fda.gov/drug/label.json?&limit=1000')
json_response = response.json()
json_response.keys()

dict_keys(['meta', 'results'])

In [35]:
json_response['meta']

{'disclaimer': 'Do not rely on openFDA to make decisions regarding medical care. While we make every effort to ensure that data is accurate, you should assume all results are unvalidated. We may limit or otherwise restrict your access to the API in line with our Terms of Service.',
 'terms': 'https://open.fda.gov/terms/',
 'license': 'https://open.fda.gov/license/',
 'last_updated': '2021-04-15',
 'results': {'skip': 0, 'limit': 1000, 'total': 185324}}

In [36]:
# Normalise the JSON data into Pandas dataframes

df = json_normalize(json_response.get('results'))
df.head()

/opt/anaconda3/envs/python-cvcourse/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


effective_time                                inactive_ingredient  \
0       20151102                     [INACTIVE INGREDIENTS Sucrose]   
1       20150109  [INGREDIENTS: TALC, POLYMETHYL METHACRYLATE, V...   
2       20180801             [INACTIVE INGREDIENTS Sucrose/Lactose]   
3       20201215                                                NaN   
4       20190124                                                NaN   

                                      purpose  \
0               [USES Boils, Abscess, Otitis]   
1                         [Purpose Sunscreen]   
2  [USES To relieve the symptoms of itching.]   
3                                         NaN   
4                                         NaN   

                       keep_out_of_reach_of_children  \
0  [Keep this and all medication out of reach of ...   
1  [Keep out of reach of children If product is s...   
2  [KEEP OUT OF REACH OF CHILDREN Keep this and a...   
3                                                NaN   
4                                                NaN   

                                            warnings  \
0  [WARNINGS This product is to be used for self-...   
1                  [Warnings For external use only.]   
2  [STOP USE AND ASK DOCTOR If symptoms persist/w...   
3  [WARNINGS NOT FOR INJECTION. Ofloxacin ophthal...   
4                                                NaN   

                                           questions  \
0  [QUESTIONS OR COMMENTS www.Rxhomeo.com | 1.888...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                           spl_product_data_elements version  \
0  [SILICEA SILICEA SILICON DIOXIDE COLLOIDAL SIL...       1   
1  [CHANTECAILLE PROTECTION NATURELLE BRONZE SPF ...       4   
2  [Mezereum DAPHNE MEZEREUM BARK SUCROSE LACTOSE...       2   
3  [Ofloxacin Ofloxacin OFLOXACIN OFLOXACIN Sodiu...       3   
4  [Naproxen Naproxen NAPROXEN NAPROXEN CROSCARME...      21   

                           dosage_and_administration  \
0  [DOSAGE Adults- Take 4 or 6 Pellets by mouth, ...   
1  [Directions Protection Naturelle SPF 46 PA+++ ...   
2  [DIRECTIONS Adults: Dissolve 3 to 5 under the ...   
3  [DOSAGE AND ADMINISTRATION The recommended dos...   
4  [2 DOSAGE AND ADMINISTRATION Use the lowest ef...   

                         pregnancy_or_breast_feeding  ...  \
0  [As with any drug, if you are pregnant, or nur...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

  package_label_principal_display_panel_table contraindications_table  \
0                                         NaN                     NaN   
1                                         NaN                     NaN   
2                                         NaN                     NaN   
3                                         NaN                     NaN   
4                                         NaN                     NaN   

  nonclinical_toxicology_table  \
0                          NaN   
1                          NaN   
2                          NaN   
3                          NaN   
4                          NaN   

  carcinogenesis_and_mutagenesis_and_impairment_of_fertility_table  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  information_for_owners_or_caregivers patient_medication_information_table  \
0                                  N

### Extracting year from 'effective_time'

In [37]:
df['year'] = pd.DatetimeIndex(df['effective_time']).year

### Extracting only the columns of interest

In [38]:
df = df[['year','spl_product_data_elements', 'openfda.route']]
df.head()

year                          spl_product_data_elements openfda.route
0  2015  [SILICEA SILICEA SILICON DIOXIDE COLLOIDAL SIL...           NaN
1  2015  [CHANTECAILLE PROTECTION NATURELLE BRONZE SPF ...           NaN
2  2018  [Mezereum DAPHNE MEZEREUM BARK SUCROSE LACTOSE...        [ORAL]
3  2020  [Ofloxacin Ofloxacin OFLOXACIN OFLOXACIN Sodiu...  [OPHTHALMIC]
4  2019  [Naproxen Naproxen NAPROXEN NAPROXEN CROSCARME...        [ORAL]

### Removing NaN values 

df.isna().sum()

In [40]:
df = df[~df['openfda.route'].isna()]

In [41]:
df.isna().sum()

year                         0
spl_product_data_elements    0
openfda.route                0
dtype: int64

### Get the number of ingredients

In [42]:
# Since each spl_product_data_element is separated from each other by commas, we split them and get the number of elements

df['number_of_ingredients'] = df['spl_product_data_elements'].apply(lambda x: len(x[0].split(',')))
df.head()

year                          spl_product_data_elements openfda.route  \
2  2018  [Mezereum DAPHNE MEZEREUM BARK SUCROSE LACTOSE...        [ORAL]   
3  2020  [Ofloxacin Ofloxacin OFLOXACIN OFLOXACIN Sodiu...  [OPHTHALMIC]   
4  2019  [Naproxen Naproxen NAPROXEN NAPROXEN CROSCARME...        [ORAL]   
5  2020  [Bactoshield CHG CHLORHEXIDINE GLUCONATE CHLOR...     [TOPICAL]   
8  2020  [Quick Action Salicylic acid SALICYLIC ACID SA...     [TOPICAL]   

   number_of_ingredients  
2                      1  
3                      1  
4                      2  
5                      1  
8                      1

### Grouping the data by year and route, averaging the number of ingredients

In [45]:
df['route'] = df['openfda.route'].apply(lambda x: x[0])
df.head()

year                          spl_product_data_elements openfda.route  \
2  2018  [Mezereum DAPHNE MEZEREUM BARK SUCROSE LACTOSE...        [ORAL]   
3  2020  [Ofloxacin Ofloxacin OFLOXACIN OFLOXACIN Sodiu...  [OPHTHALMIC]   
4  2019  [Naproxen Naproxen NAPROXEN NAPROXEN CROSCARME...        [ORAL]   
5  2020  [Bactoshield CHG CHLORHEXIDINE GLUCONATE CHLOR...     [TOPICAL]   
8  2020  [Quick Action Salicylic acid SALICYLIC ACID SA...     [TOPICAL]   

   number_of_ingredients       route  
2                      1        ORAL  
3                      1  OPHTHALMIC  
4                      2        ORAL  
5                      1     TOPICAL  
8                      1     TOPICAL

In [46]:
result2 = df.groupby(['year','route'], as_index=False).mean()

In [47]:
result2.head()

year        route  number_of_ingredients
0  2011         ORAL                    1.0
1  2015      TOPICAL                    2.0
2  2016         ORAL                    1.5
3  2016      TOPICAL                    1.0
4  2017  INTRAVENOUS                    1.0

### Customising the column names as per requirements

In [49]:
result2.columns = ['year','route','avg_number_of_ingredients']
result2

year                     route  avg_number_of_ingredients
0   2011                      ORAL                   1.000000
1   2015                   TOPICAL                   2.000000
2   2016                      ORAL                   1.500000
3   2016                   TOPICAL                   1.000000
4   2017               INTRAVENOUS                   1.000000
5   2017                      ORAL                   4.400000
6   2017                   TOPICAL                   1.875000
7   2018                    DENTAL                   1.000000
8   2018               INTRAVENOUS                   1.000000
9   2018                OPHTHALMIC                   5.000000
10  2018                      ORAL                   2.166667
11  2018                SUBLINGUAL                   7.000000
12  2018                   TOPICAL                   1.937500
13  2019                    DENTAL                   1.500000
14  2019             INTRAMUSCULAR                   3.000000
15  2019               INTRAVENOUS                   1.000000
16  2019                OPHTHALMIC                   1.000000
17  2019                      ORAL                   2.916667
18  2019                    RECTAL                   2.000000
19  2019                   TOPICAL                   2.500000
20  2019                   VAGINAL                   1.000000
21  2020                    DENTAL                   1.000000
22  2020             INTRAMUSCULAR                   1.000000
23  2020               INTRATHECAL                   1.000000
24  2020               INTRAVENOUS                   1.285714
25  2020                IRRIGATION                   1.000000
26  2020                     NASAL                   3.000000
27  2020                OPHTHALMIC                   1.333333
28  2020                      ORAL                   3.515873
29  2020                    RECTAL                   1.000000
30  2020  RESPIRATORY (INHALATION)                   1.600000
31  2020                SUBLINGUAL                   5.000000
32  2020                   TOPICAL                   1.658537
33  2020               TRANSDERMAL                   2.000000
34  2020                   VAGINAL                   1.000000
35  2021                    DENTAL                   2.000000
36  2021             INTRAMUSCULAR                   5.000000
37  2021               INTRAVENOUS                   1.000000
38  2021                      ORAL                   3.066667
39  2021  RESPIRATORY (INHALATION)                   1.000000
40  2021                   TOPICAL                   2.117647

Printing the date and time stamp after completion of the assignment

In [51]:
import datetime
now = datetime.datetime.now()
print ("Finishing date and time : ")
print (now.strftime("%Y-%m-%d %H:%M:%S"))

Finishing date and time : 
2021-04-16 13:33:04
